man/dbpack.1
.TH DBPACK 1 "12 September 1995"
.FB
.SH NAME
dbpack \- remove deleted records from a \fIFirstBase\fP database
.SH SYNOPSIS
.B dbpack
[\fB-b\fP] [\fB-y\fP]
\fIdatabase\fP [\fIdatabase\fP ...]
.SH DESCRIPTION
.I Dbpack
squeezes out all deleted records from a \fIFirstBase\fP database, leaving only the
pure, 'packed' database.
The packing is done in place.
.PP
The \fB-b\fP
flag will run the \fIFirstBase\fP tool in batch mode, suitable for shell scripts.
This mode produces no output on the terminal, thus can also be used
when running the process in the background.
The \fB-y\fP flag will run the tool at the terminal without pausing to ask
the user if existing files can be destroyed and without displaying
a 'picture' of the requested task.
.PP
Before
\fIdbpack\fP(1)
begins, the user is given a chance to abort.
.PP
.I Note:
packing a
.I FirstBase
database makes all indexes for that database obsolete and unusable.
Thus all indexes for that database must be regenerated after packing.
.PP
\fIDbclean\fP(1)
is automatically run at the completion of the packing process to
create a map file for the newly packed database.
.SH FILES
.PD 0
.TP 10
dbase
default name of database.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.ddict
\fIFirstBase\fP database dictionary file.
.PD
.SH SEE ALSO
dbclean(1), dbedit(1), dbigen(1), dbregen(1), firstbase(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br
.SH DIAGNOSTICS
\fIDbpack\fP(1)
requires a database dictionary and a database. No indexes are used.
All records are processed in database sequential order.
Dbpack will noisily do nothing if there is no work to do (i.e. no
deletions).
.sp 1
Also, \fIdbpack\fP(1) needs to be run in the same directory as the
database (or at least on the same logical device) as temporary storage is
used in the current working directory to build the new database.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbtools/dbpack.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbpack.c,v 9.1 2001/02/16 19:47:51 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char sccsid[] = "@(#)dbpack.c	8.3 06/18/00 FB";
#endif

/*
 *  dbpack.c - (dbundo.c) pack or undo deleteions from an entire
 *     Cdb fb_database.
 */

#include <fb.h>
#include <fb_vars.h>

extern char *cdb_pgm;
extern short int cdb_secure;
extern char *cdb_DBCLEAN;

static char tempname[] = {".pckXXXXXX"};	/* for temp file name */
static char fname[FB_MAXNAME];			/* for filename */
static int pack;
static void spack();
static packemin();

/*
 *  dbpack - main driver
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
   
      {
         char dname[FB_MAXNAME];
	 int j;

         fb_getargs(argc, argv, FB_NODB);
	 pack = (strcmp(cdb_pgm, "dbpack") == 0) ? 1 : 0;
	 for (j = 1; j < argc; j++)
	    if (!(argv[j][0] == '-')){
	       fb_rootname(dname, argv[j]);
	       spack(dname);
	       }
	 fb_ender();
      }

/*
 *  spack - simple pack for one fb_database.
 */

   static void spack(dname)
      char *dname;
      
      {
 	 fb_database *hp;
	 char msg[FB_MAXLINE];
	 
	 hp = fb_dballoc();
	 fb_dbargs(dname, NIL, hp);
         if (access(hp->dbase, 0) != 0)
            return;
	 fb_opendb(hp, READWRITE, FB_NOINDEX, 0);
	 if (!cdb_batchmode){
            fb_scrhdr(hp, NIL); fb_scrtime(hp);
            (pack) ? fb_scrstat("Packing") : fb_scrstat("Undoing");
	    }
         if (cdb_secure && fb_getuid() != 0)
            fb_xerror(FB_MESSAGE, "Must be FirstBase Root (UID=0) to Pack/Undo",NIL);
	 if (hp->delcnt == 0L){
	    if (cdb_batchmode || cdb_yesmode)
	       return;
	    if (fb_mustbe('y',
	          "Delete Count is 0! Still want to do this (y/n) ?",
	           cdb_t_lines, 1) == FB_ERROR)
	       return;
	    }
	 if (!cdb_batchmode && !cdb_yesmode){
	    if (pack)
	       strcpy(msg, "Ready to PACK? Are You SURE (y/n)?");
	    else
	       strcpy(msg, "Ready to UNDO? Are You SURE (y/n)?");
	    if (fb_mustbe('y', msg, cdb_t_lines, 1) == FB_ERROR)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
	 packemin(hp);
	 if (1){ /* if dbundo does work in place, only pack needs cleaning */
	 	 /* for now, always do cleaning -- makes indexes obsolete */
	    fb_rootname(fname, hp->dbase);
	    if (!cdb_batchmode){
	       fb_move(4,1), fb_clrtobot();
	       fb_scrstat("Cleaning");
	       sprintf(msg, "Cleaning %s", fname);
	       fb_fmessage(msg);
	       }
	    sprintf(msg, "%s -b %s", cdb_DBCLEAN, fname);
	    fb_system(msg, FB_WITHROOT);
	    }
         fb_closedb(hp);
      }
      
/* 
 *  packemin - pack or undo a fb_database 
 */
 
   static packemin(hp)
      fb_database *hp;
      
      {
         long rec, n;
	 int fd, size;
	 char p[FB_SEQSIZE+2];

         n = 0L;
	 if (!cdb_batchmode){
	    FB_XWAIT();
	    fb_gcounter(n);
	    }
	 close(mkstemp(tempname));
	 fd = fb_mustopen(tempname, READWRITE);
	 if (pack)				/* pack demands new dbase */
	    fb_putseq(fd);
	 else{					/* post undo indexes are ok */
	    sprintf(p, "%04d", fb_getseq(hp->fd));
	    lseek(fd, FB_SEQSTART, 0);
	    write(fd, p, FB_SEQSIZE);
	    }
         fb_puthead(fd, 0L, 0L);
         if (cdb_secure)
            if (fb_putmode(fd, fb_getuid(), fb_getgid(), "666") == FB_ERROR)
               fb_xerror(FB_IO_ERROR, hp->dbase, NIL);
	 for (rec = 1L; rec <= hp->reccnt; rec++){
	    if ((size = fb_getrec(rec, hp)) == FB_ERROR)
	       fb_xerror(FB_FATAL_GETREC, hp->dbase, (char *) &rec);
	    if (FB_ISDELETED(hp)){
	       if (pack)
	          continue;
	       else
	          hp->kp[hp->nfields]->fld[0] = FB_BLANK;
	       }
	    if (write(fd, hp->orec, size) != size)
	       fb_xerror(FB_WRITE_ERROR, cdb_pgm, hp->dbase);
	    ++n;
	    if (!cdb_batchmode)
	       fb_gcounter(n);
	    }
	 if (fb_puthead(fd, n, 0L) == FB_ERROR)
	    fb_xerror(FB_IO_ERROR, SYSMSG[S_BAD_HEADER], hp->dbase);
	 close(fd);
	 unlink(hp->dbase);
	 if (link(tempname,hp->dbase) < 0)
	    if (fb_copyfile(tempname,hp->dbase) < 0)
	       fb_xerror(FB_EXEC_FAILURE, "Could not link/copy: ", tempname);
	 unlink(tempname);
      }

/*  
 *  usage message 
 */
 
   usage()
      {
         fb_xerror(FB_MESSAGE, "usage: dbpack/dbundo [-b] -[y] dbase", NIL);
      }
